xquant/keras/keras_report_utils.py

```diff
@@ -0,0 +1,380 @@
+
+            metrics_to_compute += list(custom_metrics_intermediate.keys())
+
+        float_name2quant_name = self.get_float_to_quantized_compare_points(float_model=float_model,
+                                                                           quantized_model=quantized_model)
+
+        results = {q_layer: [] for q_layer in float_name2quant_name.values()}
+
+        for x in dataset():
+            quant_activations = get_activations(quantized_model, list(float_name2quant_name.values()), x)
+            float_activations = get_activations(float_model, list(float_name2quant_name.keys()), x)
+
+            for float_layer, quant_layer in float_name2quant_name.items():
+                float_activation = float_activations[float_layer]
+                quant_activation = quant_activations[quant_layer]
+
+                results[quant_layer].append(
+                    self.compute_metrics((float_activation, quant_activation), custom_metrics_intermediate))
+
+        aggregated_metrics = {}
+        for layer_name, layer_metrics in results.items():
+            combined_dict = {}
+            for item in layer_metrics:
+                for key, value in item.items():
+                    if key not in combined_dict:
+                        combined_dict[key] = []
+                    combined_dict[key].append(value)
+            for k, v in combined_dict.items():
+                combined_dict[k] = np.mean(v)
+            aggregated_metrics[layer_name] = combined_dict
+
+        return aggregated_metrics
+
+    def get_default_metrics(self) -> Dict[str, Callable]:
+        """
+        Retrieve the default metrics for evaluation.
+
+        Returns:
+            Dict[str, Callable]: A dictionary of default metric functions.
+        """
+
+        def compute_mse(f_pred: np.ndarray, q_pred: np.ndarray) -> float:
+            mse = tf.keras.losses.MeanSquaredError()(f_pred, q_pred)
+            return float(mse.numpy())
+
+        def compute_cs(f_pred: np.ndarray, q_pred: np.ndarray) -> float:
+            cs = tf.keras.losses.CosineSimilarity()(f_pred.flatten(), q_pred.flatten())
+            return float(cs.numpy())
+
+        def compute_sqnr(f_pred: np.ndarray, q_pred: np.ndarray) -> float:
+            signal_power = tf.reduce_mean(tf.square(f_pred))
+            noise_power = tf.reduce_mean(tf.square(f_pred - q_pred))
+            sqnr = signal_power / noise_power
+            return float(sqnr.numpy())
+
+        return {MSE_METRIC_NAME: compute_mse,
+                CS_METRIC_NAME: compute_cs,
+                SQNR_METRIC_NAME: compute_sqnr}
+
+    def create_float_folded_model(self,
+                                  float_model: keras.Model,
+                                  representative_dataset: Any,
+                                  ) -> keras.Model:
+        """
+        Create a folded version of the floating-point model.
```

Comment:
give some more explanation of what folded model is.
Is it only a model with folded batchnorms? are there any other substitutions/changes to the model structure applied here?

Reply:
It applies all graph substitutions that MCT applies so the structure is identical (except for things like SNC that irrelevant for the float structure).